File: Prisoner.java
abstract class Prisoner {

    protected String strategy;
    protected int score;
    //0 is defect, 1 is cooperate
    protected int lastMove;
    //Counts how many of each move was done in this bout, using the move as an index
    protected int[] moves;
    protected Prisoner opponent;

    public Prisoner(String strategy) {
        //Set up the move count array
        this.strategy = strategy;
        //Set up moves and lastMove
        reset();
    }

    public String toString() {
        //String representation of the prisoner
        return strategy + " Prisoner: Score - " + score;
    }

    public String getStrategy() {
        return strategy;
    }

    public int getScore() {
        // Return the score for this prisoner
        return score;
    }

    public int getMajority() {
        //Returns the move this prisoner made more of, or -1 if tie
        if (moves[0] > moves[1]) {
            return 0;
        }
        else if (moves[0] < moves[1]) {
            return 1;
        }
        else {
            return -1;
        }
    }

    public int getLastMove() {
        //Return the last move made by this prisoner
        return lastMove;
    }

    public void setLastMove(int lastMove) {
        //Update this prisoner's last move
        this.lastMove = lastMove;
        //Increment the count of the moves played
        this.moves[lastMove] ++;
    }

    public void updateScore(int score) {
        //Update this prisoner's score
        this.score += score;
    }

    public void reset() {
        //Reset the prisoner's move count in between bouts
        moves = new int[2];
        //Set the last move to -1 to indicate that no last move was made
        lastMove = -1;
    }

    public void setOpponent(Prisoner opponent) {
        //Update the opponent variable to the current opponent
        this.opponent = opponent;
    }

    //Abstract play method, to be overwritten in the appropriate subclasses
    public abstract int play();

    public abstract Prisoner clone();

    public int hashCode() {
        return strategy.hashCode();
    }

    public boolean equals(Object obj) {
        //For the sake of the HashMap, check the equality by strategy
        if (this == obj) {
            return true;
        }
        if (obj == null || this.getClass() != obj.getClass()) {
            return false;
        }
        Prisoner p = (Prisoner) obj;
        return (p.strategy.equals(this.strategy));
    }
}

File: prisoner/Prisoner.kt
package de.freyama.prisoners_dilemma.prisoner

/*
 * This is an abstract class that all Prisoner Strategy classes will follow.
 * There are two moves that can be made in Prisoner's Dilemma; Defect (0) or Cooperate (1).
 */
abstract class Prisoner {
    // Instance Variables

    /**
     * The last move that the Strategy played.
     * This is necessary in two situations; either the strategy needs to know the last move it played
     * or it needs to know the last move its opponent played.
     */
    private var lastMove: Boolean? = null
    /**
     * Keeps track of how many of each move was played by the strategy so far.
     * `.first` is the count of defects, and `.second` is the count of cooperates.
     */
    private var moveCounts: Pair<Int, Int> = Pair(0, 0)
    /**
     * Maintains a reference to the current opponent being faced.
     * Some Prisoners remember moves their opponents use against them to determine moves to make.
     */
    protected var opponent: Prisoner? = null
    /**
     * The current score that the strategy has received
     */
    protected var score: Int = 0
    /**
     * Strategy name
     */
    protected val strategy: String

    constructor(strategy: String) {
        this.strategy = strategy
        // Reset the current state of the Prisoner
        this.reset()
    }

    // Overrides

    /**
     * Convert the Prisoner class into a String
     */
    override fun toString(): String = "$strategy Prisoner: Score - $score"

    fun equals(other: Prisoner) = this.strategy.equals(other.strategy)

    // Abstract Methods

    /**
     * Make a move.
     * Returning True symbolises cooperation, False symbolises defection.
     */
    abstract fun play(): Boolean

    // Concrete Methods

    /**
     * Get the previously made move by the Prisoner
     */
    fun getLastMove(): Boolean? {
        return this.lastMove
    }

    /**
     * Get the move played most by the opponent.
     * null will be returned when the opponent has no majority.
     */
    fun getMajority(): Boolean? {
        if (this.moveCounts.first == this.moveCounts.second) {
            return null
        }
        return this.moveCounts.second > this.moveCounts.first
    }

    /**
     * Reset the state of the Prisoner, by deleting the move counts and last move values.
     * This function is called after two prisoners finish playing against eachother, as these details are recorded for a pair of Prisoners.
     */
    private fun reset() {
        // Reset the moveCounts
        this.moveCounts = Pair(0, 0)
        // Set the lastMove to null to indicate there was no last move
        this.lastMove = null
    }
}
